File: internal/httpapi/templates.go
package httpapi

import (
	"net/http"

	"github.com/go-freya/freya/services/notification/internal/notify"
)

type templateBody struct {
	Name      string   `json:"name"`
	ChannelID string   `json:"channel_id"`
	Subject   string   `json:"subject"`
	Body      string   `json:"body"`
	Variables []string `json:"variables"`
	IsDefault bool     `json:"is_default"`
}

func (b templateBody) input() notify.TemplateInput {
	return notify.TemplateInput{Name: b.Name, ChannelID: b.ChannelID, Subject: b.Subject, Body: b.Body, Variables: b.Variables, IsDefault: b.IsDefault}
}

// MaxTemplateBody bounds template create/update/preview bodies (256 KiB body + slack).
const MaxTemplateBody = 300 << 10

// RegisterTemplates mounts the template routes (contracts §templates).
func (s *Server) RegisterTemplates(d NotifyDeps) {
	s.MustHandle("GET", Prefix+"/templates", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		q := r.URL.Query()
		var channelID *string
		if c := q.Get("channel_id"); c != "" {
			channelID = &c
		}
		items, next, err := d.Templates.List(r.Context(), subj, channelID, q.Get("q"), q.Get("cursor"), limitParam(r))
		if err != nil {
			s.fail(w, r, domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{"items": items, "next_cursor": next})
	})
	s.MustHandle("POST", Prefix+"/templates", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in templateBody
		if err := DecodeJSON(r, &in, MaxTemplateBody); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Templates.Create(r.Context(), subj, in.input())
		if err != nil {
			s.fail(w, r, domainError(err))
			return
		}
		WriteJSON(w, http.StatusCreated, out)
	})
	s.MustHandle("GET", Prefix+"/templates/{id}", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Templates.Get(r.Context(), subj, r.PathValue("id"))
		if err != nil {
			s.fail(w, r, readError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("PUT", Prefix+"/templates/{id}", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in templateBody
		if err := DecodeJSON(r, &in, MaxTemplateBody); err != nil {
			Fail(w, r, nil, err)
			return
		}
		out, err := d.Templates.Update(r.Context(), subj, r.PathValue("id"), in.input())
		if err != nil {
			s.fail(w, r, domainError(err))
			return
		}
		WriteJSON(w, http.StatusOK, out)
	})
	s.MustHandle("POST", Prefix+"/templates/{id}/remove", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		if err := d.Templates.Delete(r.Context(), subj, r.PathValue("id")); err != nil {
			s.fail(w, r, domainError(err))
			return
		}
		w.WriteHeader(http.StatusNoContent)
	})
	s.MustHandle("POST", Prefix+"/templates/preview", func(w http.ResponseWriter, r *http.Request) {
		subj, err := subjects(r)
		if err != nil {
			Fail(w, r, nil, err)
			return
		}
		var in struct {
			TemplateID  string            `json:"template_id"`
			ChannelType string            `json:"channel_type"`
			Subject     string            `json:"subject"`
			Body        string            `json:"body"`
			Variables   []string          `json:"variables"`
			Values      map[string]string `json:"values"`
		}
		if err := DecodeJSON(r, &in, MaxTemplateBody); err != nil {
			Fail(w, r, nil, err)
			return
		}
		subject, body, err := d.Templates.Preview(r.Context(), subj, notify.PreviewInput{TemplateID: in.TemplateID, ChannelType: in.ChannelType, Subject: in.Subject, Body: in.Body, Variables: in.Variables, Values: in.Values})
		if err != nil {
			s.fail(w, r, readError(err))
			return
		}
		WriteJSON(w, http.StatusOK, map[string]string{"rendered_subject": subject, "rendered_body": body})
	})
}

File: ui/src/stores/templates.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '@/api/client'
import type { Page, Template, TemplateInput } from '@/api/types'

export interface Preview {
  rendered_subject: string
  rendered_body: string
}

export interface PreviewInput {
  template_id?: string | undefined
  channel_type?: string | undefined
  subject: string
  body: string
  variables?: string[]
  values?: Record<string, string>
}

export const useTemplates = defineStore('notification-templates', () => {
  const items = ref<Template[]>([])
  const next = ref<string | undefined>()
  const loading = ref(false)
  const error = ref('')

  async function list(channelId?: string, q?: string, cursor?: string): Promise<void> {
    loading.value = true
    error.value = ''
    try {
      const page = await api<Page<Template>>('GET', 'templates', undefined, { query: { channel_id: channelId, q, cursor, limit: 50 } })
      items.value = cursor ? [...items.value, ...page.items] : page.items
      next.value = page.next_cursor
    } catch (e) {
      error.value = (e as Error).message
    } finally {
      loading.value = false
    }
  }

  async function get(id: string): Promise<Template> {
    return api<Template>('GET', 'templates/' + id)
  }

  async function create(input: TemplateInput): Promise<Template> {
    const t = await api<Template>('POST', 'templates', input)
    items.value = [t, ...items.value]
    return t
  }

  async function update(id: string, input: TemplateInput): Promise<Template> {
    const t = await api<Template>('PUT', 'templates/' + id, input)
    items.value = items.value.map((x) => (x.id === id ? t : x))
    if (t.is_default) items.value = items.value.map((x) => (x.id !== id && x.channel_id === t.channel_id ? { ...x, is_default: false } : x))
    return t
  }

  async function remove(id: string): Promise<void> {
    await api('POST', 'templates/' + id + '/remove')
    items.value = items.value.filter((x) => x.id !== id)
  }

  async function preview(input: PreviewInput): Promise<Preview> {
    return api<Preview>('POST', 'templates/preview', input)
  }

  return { items, next, loading, error, list, get, create, update, remove, preview }
})
